lib/chrono_model/time_machine/history_model.rb

```diff
@@ -219,7 +219,27 @@ def valid_from
       def valid_to
         validity.end if validity.end.is_a?(Time)
       end
-      alias as_of_time valid_to
+
+      # For historical records, as_of_time should be within the validity period
+      # to ensure association queries return records that existed during this
+      # record's validity, not records that became valid at the boundary time.
+      #
+      # When objects are updated in the same transaction, they get the same valid_to
+      # timestamp, creating an edge case where association queries return the wrong version.
+      # PostgreSQL ranges are half-open [start, end) by default.
+      def as_of_time
+        # For current objects with infinite valid_to, use the time as-is
+        # For historical objects with finite valid_to, subtract 1 unit at precision 6 to avoid edge case
+        if valid_to&.infinite?
+          valid_to  # Keep infinite time as-is
+        elsif valid_to.nil?
+          nil       # Handle nil case for backward compatibility
+        else
+          # Subtract 1 unit at precision 6 to get "just before" the boundary
+          # This ensures we get the historical version, not the current one
+          valid_to - ChronoModel::TIMESTAMP_PRECISION
+        end
```

Comment:
I'm still not convinced

1. In the comment, Add a reference to `ifad/chronomodel#283` (include `ifad/chronomodel#283` as is)
2. In the method, Early return if it is nil (guard clause, please also run rubocop after you change something)

At this point we can have either:
- time value (is not nil, does not respond to infinite) So i guess that we can subtract one
- other value, either infinity or nil (actually just "nil", but let's plan for the future)

```
if valid_to.is_a?(Time)
  # subtract
else
  valid_to
end
```

Should be enough.

Please do not trust me. Check. And do not use adverbs (eg: "you are correct" is good, "you are absolutely correct" is not)